src/core/provider.cc

```diff
@@ -84,7 +88,10 @@ AllocatedSystemMemory::AllocatedSystemMemory(
       }
     } else {
 #ifdef TRTIS_ENABLE_GPU
-      cudaError_t err = cudaMalloc((void**)&buffer_, byte_size);
+      auto err = cudaSetDevice(memory_type_id_);
```

Comment:
Are callers of AllocatedSystemMemory all ok with the cuda device changing? I think so but we should check to make sure...